### ISSUE TITLE
Weird behavior cloning repos on the same filesystem

### ISSUE THREAD

**Issue by abraithwaite** [github/git-lfs#195]
I tried to clone a repository to another directory on the local filesystem and encountered this unexpected behavior.

```shell
abraithwaite at jessie in /tmp
$ git clone asdf qwer
Cloning into 'qwer'...
done.
Downloading asdf.zip (0 B)
Username for '://':
Password for '://':
Error accessing media: asdf.zip (e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855)

Errors logged to /tmp/qwer/.git/lfs/objects/logs/20150408T203322.609792813.log.
Use `git lfs logs last` to view the log.

abraithwaite at jessie in /tmp
$ git lfs logs last
No logs to show

$ cat  /tmp/qwer/.git/lfs/objects/logs/20150408T203322.609792813.log
> git-lfs smudge asdf.zip
LocalWorkingDir=/tmp/qwer
LocalGitDir=/tmp/qwer/.git
LocalMediaDir=/tmp/qwer/.git/lfs/objects
TempDir=/tmp/qwer/.git/lfs/tmp
GIT_DIR=.git

Error accessing media: asdf.zip (e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855)

Error downloading /tmp/qwer/.git/lfs/objects/e3/b0/e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855.
Get %3Cunknown%3E/objects/e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855: unsupported protocol scheme ""
Response:Authorization=--
Response:Accept=application/vnd.git-lfs+json; charset-utf-8
Status=
Endpoint=<unknown>
URL=GET %3Cunknown%3E/objects/e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
Response:User-Agent=git-lfs/0.5.0 (GitHub; linux amd64; git 2.1.4; go 1.4.2-cf1)
goroutine 1 [running]:
github.com/github/git-lfs/lfs.Stack(0x0, 0x0, 0x0)
        /home/abraithwaite/Projects/git-lfs/.vendor/src/github.com/github/git-lfs/lfs/errors.go:85 +0x8d
github.com/github/git-lfs/lfs.Errorf(0x7f902835b238, 0xc20800bbf0, 0x7d1f90, 0xf, 0xc208077740, 0x2, 0x2, 0x1)
        /home/abraithwaite/Projects/git-lfs/.vendor/src/github.com/github/git-lfs/lfs/errors.go:29 +0x5e
github.com/github/git-lfs/lfs.doHttpRequest(0xc208034d00, 0xc20800bb60, 0x8, 0xc2080104f1)
        /home/abraithwaite/Projects/git-lfs/.vendor/src/github.com/github/git-lfs/lfs/client.go:231 +0x192
github.com/github/git-lfs/lfs.doApiRequestWithRedirects(0xc208034d00, 0xc20800bb60, 0xc208077898, 0x0, 0x4, 0xc208077860, 0xc208070000)
        /home/abraithwaite/Projects/git-lfs/.vendor/src/github.com/github/git-lfs/lfs/client.go:252 +0x40
github.com/github/git-lfs/lfs.doApiRequest(0xc208034d00, 0xc20800bb60, 0xc2080104f1, 0x40, 0xc208034d00)
        /home/abraithwaite/Projects/git-lfs/.vendor/src/github.com/github/git-lfs/lfs/client.go:287 +0x85
github.com/github/git-lfs/lfs.Download(0xc2080104f1, 0x40, 0x0, 0x0, 0xc208077a30, 0x2)
        /home/abraithwaite/Projects/git-lfs/.vendor/src/github.com/github/git-lfs/lfs/client.go:102 +0xca
github.com/github/git-lfs/pointer.downloadFile(0x7f9028359b70, 0xc208038008, 0xc208056080, 0x7ffd01c76630, 0x8, 0xc2080104d0, 0x61, 0x0, 0x0)
        /home/abraithwaite/Projects/git-lfs/.vendor/src/github.com/github/git-lfs/pointer/smudge.go:35 +0x233
github.com/github/git-lfs/pointer.Smudge(0x7f9028359b70, 0xc208038008, 0xc208056080, 0x7ffd01c76630, 0x8, 0x0, 0x0, 0x0)
        /home/abraithwaite/Projects/git-lfs/.vendor/src/github.com/github/git-lfs/pointer/smudge.go:21 +0x2f1
github.com/github/git-lfs/pointer.(*Pointer).Smudge(0xc208056080, 0x7f9028359b70, 0xc208038008, 0x7ffd01c76630, 0x8, 0x0, 0x0, 0x0)
        /home/abraithwaite/Projects/git-lfs/.vendor/src/github.com/github/git-lfs/pointer/pointer.go:43 +0x6f
github.com/github/git-lfs/commands.smudgeCommand(0x9b7080, 0xc20802a4a0, 0x1, 0x1)
        /home/abraithwaite/Projects/git-lfs/.vendor/src/github.com/github/git-lfs/commands/command_smudge.go:59 +0x9b4
github.com/spf13/cobra.(*Command).execute(0x9b7080, 0xc20802a490, 0x1, 0x1, 0x0, 0x0)
        /home/abraithwaite/Projects/git-lfs/.vendor/src/github.com/spf13/cobra/command.go:355 +0x189
github.com/spf13/cobra.(*Command).findAndExecute(0x9b7680, 0xc20800a010, 0x2, 0x2, 0x0, 0x0)
        /home/abraithwaite/Projects/git-lfs/.vendor/src/github.com/spf13/cobra/command.go:334 +0xd4
github.com/spf13/cobra.(*Command).Execute(0x9b7680, 0x0, 0x0)
        /home/abraithwaite/Projects/git-lfs/.vendor/src/github.com/spf13/cobra/command.go:408 +0x8ad
github.com/github/git-lfs/commands.Run()
        /home/abraithwaite/Projects/git-lfs/.vendor/src/github.com/github/git-lfs/commands/commands.go:80 +0x2a
main.main()
        /home/abraithwaite/Projects/git-lfs/cmd/git-lfs.go:8 +0x1b
```

**Comment by tclem** [comment]
Thanks for the report @abraithwaite!

At the moment, you'd have to have something at the remote (in this case on your local machine) that speaks the Git LFS [server API](https://github.com/github/git-lfs/blob/master/docs/spec.md). Can you explain your use case here in a bit more detail? It would help us understand the importance of a feature like this. Thanks!

**Comment by abraithwaite** [comment]
I was trying to clone a local repository from one directory to another (probably not too uncommon of an operation) when I was prompted for the username and password for the repository.

I haven't set one up clearly, and I'm still trying to find what calls this, but I would expect the behavior to be cleaner handling of a situation like this.  I'm guessing this is done by the smudge-filter?

Yep: https://github.com/github/git-lfs/blob/master/pointer/smudge.go#L23

I'm looking into why it wasn't trying to get it from the local filesystem now.

**Comment by gitfoxi** [comment]
+1

I think local clone is a critical feature. Even if git-lfs doesn't account for it yet, we're really just talking about copying files.

One question is, does it make more sense to copy the large files or simply use symbolic links? Symbolic links would be attractive for saving disk space. But how do you guarantee the linked file never changes or disappears? You could make large files read-only after commit. It still wouldn't keep someone from moving or removing an upstream repo. So I think it's better to just copy the large files unless that question can be answered.

**Comment by tclem** [comment]
:cool: Thanks for explaining. We should definitely just fix this then and I'd probably vote to just copy the large files as a starting place.

**Comment by technoweenie** [comment]
> I was trying to clone a local repository from one directory to another (probably not too uncommon of an operation)

What's a real use case for this?  Honestly the only time I've ever done this is in our integration test suites for GitHub itself.  I think it'd be nice to support local files, but I don't see it as a priority.  If you have a valid use case though, I'm all ears.

I definitely want to fix the explosion, though.  Perhaps just bail early on any Git LFS calls if the remote url isn't SSH/https/git.  

> One question is, does it make more sense to copy the large files or simply use symbolic links? Symbolic links would be attractive for saving disk space. 

Can we make symlinks work with smudge/clean filters?  I'd be up for that.

> But how do you guarantee the linked file never changes or disappears? You could make large files read-only after commit.

Good question :)

**Comment by abraithwaite** [comment]
> What's a real use case for this?

Being able to clone to anything you can mount is a pretty neat feature of git.  e.g. Clone to a flash drive for a friend, clone to a local fileserver, cloning for backups, etc.  I think it's especially relevant for large repositories where it takes a long time to download binaries over the internet. :+1:

**Comment by ddanier** [comment]
Local cloning currently breaks heavily as the files inside the new clone will never get filled with the real file contents after the initial clone. This is caused by the new clone not having the correct URL set up. So the pointer files will never get resolved.

The only workaround so far:
1. git init path/to/local/clone
2. cd path/to/local/clone
3. git config lfs.url https://domain/path/to/lfs
4. git remote add origin /path/to/local/origin
5. git pull origin master

This way the LFS URL is available before the new objects get pulled.

Anyway: I think git-lfs should get an additional command to refetch broken files. This may be handy in more use cases. For example such a state could be caused by a network issue. Should I open a new ticket for this?

**Comment by technoweenie** [comment]
> I think git-lfs should get an additional command to refetch broken files. 

This is number 4 in the [roadmap](https://github.com/github/git-lfs/blob/master/ROADMAP.md) :)

**Comment by technoweenie** [comment]
Closing this. Git LFS has improved a lot in the last year, with better commands like `git lfs pull` and `git lfs clone`.
